data-prepper-plugins/otel-proto-common/src/main/java/org/opensearch/dataprepper/plugins/otel/codec/OTelLogsProtoBufDecoder.java

```diff
@@ -0,0 +1,80 @@
+
+import io.opentelemetry.proto.collector.logs.v1.ExportLogsServiceRequest;
+import org.opensearch.dataprepper.model.codec.ByteDecoder;
+import org.opensearch.dataprepper.model.record.Record;
+
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.log.OpenTelemetryLog;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+ 
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.util.function.Consumer;
+import java.time.Instant;
+ 
+public class OTelLogsProtoBufDecoder implements ByteDecoder {
+    private static final Logger LOG = LoggerFactory.getLogger(OTelLogsProtoBufDecoder.class);
+    private final OTelProtoCodec.OTelProtoDecoder otelProtoDecoder;
+    private final boolean lengthPrefixedEncoding;
+    
+    public OTelLogsProtoBufDecoder(boolean lengthPrefixedEncoding) {
+        otelProtoDecoder = new OTelProtoCodec.OTelProtoDecoder();
+        this.lengthPrefixedEncoding = lengthPrefixedEncoding;
+    }
+
+    public void parse(InputStream inputStream, Instant timeReceivedMs, Consumer<Record<Event>> eventConsumer) throws IOException {
+        Reader reader = new InputStreamReader(inputStream);
+        ExportLogsServiceRequest.Builder builder = ExportLogsServiceRequest.newBuilder();
+        // As per the implementation of AWSS3 exporter in "otlp_proto" format at
+        // https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/awss3exporter
+        // Each request is written in a separate S3 object, so, no legth preceeding the actual data
+        // Same with Kafka exporter too. Each message is written as a separate message to Kafka
+        if (!lengthPrefixedEncoding) {
+            try {
+                byte[] buffer = inputStream.readAllBytes();
+                ExportLogsServiceRequest request = ExportLogsServiceRequest.parseFrom(buffer);
+                List<OpenTelemetryLog> logs = otelProtoDecoder.parseExportLogsServiceRequest(request, timeReceivedMs);
+                for (OpenTelemetryLog log: logs) {
+                    eventConsumer.accept(new Record<>(log));
+                }
+            } catch (Exception e) {
+                LOG.warn("Failed to parse Log Request");
+            }
+            return;
+        }
+        // As per the implementation of File exporter in "proto" format at
+        // https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/fileexporter/file_writer.go
+        // The file input stream should have length, followed by the request of "length" bytes
+        byte[] lenBytes = new byte[4];
+        while (inputStream.read(lenBytes, 0, 4) == 4) {
+            ByteBuffer lengthBuffer = ByteBuffer.wrap(lenBytes);
+            int len = lengthBuffer.getInt();
+            byte[] buf = new byte[len];
+            if (inputStream.read(buf, 0, len) != len) {
```

Comment:
Do you think, there might be a more efficient way with channels and byte buffers avoiding the materialisation of large arrays in memory?

Reply:
I am not sure I understand. Could you provide an example?

Reply:
This would be along the lines of @dlvenable comment: https://github.com/opensearch-project/data-prepper/pull/5421#discussion_r1946708731

Reply:
I found, that you can hand the input stream to the parse method directly. This would avoid creating large arrays.

Reply:
@KarstenSchnitter that's great. I did not know that inputstream can be sent directly. But it can only be used in the `readAllBytes` case, right? How can I use it here? It wouldn't know that the data is in length followed by request format, right?

Reply:
@kkondaka, I would assume, you can just hand over the inputstream and it will read until the object is complete. But I am not sure about that.